### ISSUE TITLE
MapTileModuleProviderBase.loadMapTileAsync() causes ANR

### ISSUE THREAD

**Issue by Vandalko** [osmdroid/osmdroid#757]
I don't have exact steps to reproduce right now - just running 6.0.0 version and trying to display a few markers. It does not occur each time.

```
Long monitor contention event with owner method=void org.osmdroid.tileprovider.modules.MapTileModuleProviderBase.loadMapTileAsync(org.osmdroid.tileprovider.MapTileRequestState) from MapTileModuleProviderBase.java:135 waiters=4 for 379ms
```

Here is the full trace:
```
"main" prio=5 tid=1 Blocked
  | group="main" sCount=1 dsCount=0 obj=0x74fc8fa0 self=0xb4d76500
  | sysTid=13815 nice=0 cgrp=default sched=0/0 handle=0xb6f85b34
  | state=S schedstat=( 0 0 0 ) utm=179 stm=51 core=2 HZ=100
  | stack=0xbe298000-0xbe29a000 stackSize=8MB
  | held mutexes=
  at org.osmdroid.tileprovider.modules.MapTileModuleProviderBase.loadMapTileAsync(MapTileModuleProviderBase.java:135)
  - waiting to lock <0x0730f43e> (a java.lang.Object)
  at org.osmdroid.tileprovider.MapTileProviderArray.mapTileRequestFailed(MapTileProviderArray.java:152)
  at org.osmdroid.tileprovider.modules.MapTileModuleProviderBase$1.removeEldestEntry(MapTileModuleProviderBase.java:122)
  at java.util.LinkedHashMap.addNewEntry(LinkedHashMap.java:189)
  at java.util.HashMap.put(HashMap.java:403)
  at org.osmdroid.tileprovider.modules.MapTileModuleProviderBase.loadMapTileAsync(MapTileModuleProviderBase.java:147)
  - locked <0x0037129f> (a java.lang.Object)
  at org.osmdroid.tileprovider.MapTileProviderArray.mapTileRequestFailed(MapTileProviderArray.java:152)
  at org.osmdroid.tileprovider.modules.MapTileModuleProviderBase$1.removeEldestEntry(MapTileModuleProviderBase.java:122)
  at java.util.LinkedHashMap.addNewEntry(LinkedHashMap.java:189)
  at java.util.HashMap.put(HashMap.java:403)
  at org.osmdroid.tileprovider.modules.MapTileModuleProviderBase.loadMapTileAsync(MapTileModuleProviderBase.java:147)
  - locked <0x05198fec> (a java.lang.Object)
  at org.osmdroid.tileprovider.MapTileProviderArray.mapTileRequestFailed(MapTileProviderArray.java:152)
  at org.osmdroid.tileprovider.modules.MapTileModuleProviderBase$1.removeEldestEntry(MapTileModuleProviderBase.java:122)
  at java.util.LinkedHashMap.addNewEntry(LinkedHashMap.java:189)
  at java.util.HashMap.put(HashMap.java:403)
  at org.osmdroid.tileprovider.modules.MapTileModuleProviderBase.loadMapTileAsync(MapTileModuleProviderBase.java:147)
  - locked <0x02445fb5> (a java.lang.Object)
  at org.osmdroid.tileprovider.MapTileProviderArray.getMapTile(MapTileProviderArray.java:132)
  at org.osmdroid.views.overlay.TilesOverlay$OverlayTileLooper.handleTile(TilesOverlay.java:221)
  at org.osmdroid.util.TileLooper.loop(TileLooper.java:29)
  at org.osmdroid.views.overlay.TilesOverlay$OverlayTileLooper.loop(TilesOverlay.java:208)
  at org.osmdroid.views.overlay.TilesOverlay.drawTiles(TilesOverlay.java:196)
  at org.osmdroid.views.overlay.TilesOverlay.draw(TilesOverlay.java:185)
  at org.osmdroid.views.overlay.DefaultOverlayManager.onDraw(DefaultOverlayManager.java:125)
  at org.osmdroid.views.MapView.dispatchDraw(MapView.java:1108)
  at android.view.View.updateDisplayListIfDirty(View.java:15175)
  at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:3593)
  at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:3573)
  at android.view.View.updateDisplayListIfDirty(View.java:15140)
  at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:3593)
  at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:3573)
  at android.view.View.updateDisplayListIfDirty(View.java:15140)
  at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:3593)
  at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:3573)
  at android.view.View.updateDisplayListIfDirty(View.java:15140)
  at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:3593)
  at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:3573)
  at android.view.View.updateDisplayListIfDirty(View.java:15140)
  at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:3593)
  at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:3573)
  at android.view.View.updateDisplayListIfDirty(View.java:15140)
  at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:3593)
  at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:3573)
  at android.view.View.updateDisplayListIfDirty(View.java:15140)
  at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:3593)
  at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:3573)
  at android.view.View.updateDisplayListIfDirty(View.java:15140)
  at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:3593)
  at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:3573)
  at android.view.View.updateDisplayListIfDirty(View.java:15140)
  at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:3593)
  at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:3573)
  at android.view.View.updateDisplayListIfDirty(View.java:15140)
  at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:3593)
  at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:3573)
  at android.view.View.updateDisplayListIfDirty(View.java:15140)
  at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:281)
  at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:287)
  at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:322)
  at android.view.ViewRootImpl.draw(ViewRootImpl.java:2615)
  at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:2434)
  at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2067)
  at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1107)
  at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6013)
  at android.view.Choreographer$CallbackRecord.run(Choreographer.java:858)
  at android.view.Choreographer.doCallbacks(Choreographer.java:670)
  at android.view.Choreographer.doFrame(Choreographer.java:606)
  at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:844)
  at android.os.Handler.handleCallback(Handler.java:739)
  at android.os.Handler.dispatchMessage(Handler.java:95)
  at android.os.Looper.loop(Looper.java:148)
  at android.app.ActivityThread.main(ActivityThread.java:5417)
  at java.lang.reflect.Method.invoke!(Native method)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

**Comment by Vandalko** [comment]
So, tile is added to the queue in `MapTileModuleProviderBase.loadMapTileAsync()`
```
			// this will put the tile in the queue, or move it to the front of
			// the queue if it's already present
			mPending.put(pState.getMapTile(), pState);
```

`mPending` map has limited size, with `removeEldestLogic` logic:
```
					if (result != null) {
						MapTileRequestState state = mPending.get(result);
						removeTileFromQueues(result);
						state.getCallback().mapTileRequestFailed(state);
					}
```

Looking at `MapTileProviderArray.mapTileRequestFailed`:
```
	@Override
	public void mapTileRequestFailed(final MapTileRequestState aState) {
		final MapTileModuleProviderBase nextProvider = findNextAppropriateProvider(aState);
		if (nextProvider != null) {
			nextProvider.loadMapTileAsync(aState);
		} else {
			synchronized (mWorking) {
				mWorking.remove(aState.getMapTile());
			}
			super.mapTileRequestFailed(aState);
		}
	}
```


So ANR happen during never-ending cycle in chain `loadMapTileAsync` -> `removeEldestLogic` -> `mapTileRequestFailed` -> `loadMapTileAsync`

I wonder why it happen if I have a single provider... Is it retry logic ?

**Comment by monsieurtanuki** [comment]
@Vandalko Nice bug description!
A way to solve that would be to "erase" the `aState` in `eraseEldestEntry` before calling `mapTileRequestFailed`:
* the meaning of that would be "in the current context, it's an old map tile I don't need anymore"
* in my mind erasing a `MapTileRequestState` is something like emptying its providers queue (new method easy to implement). Therefore `null` will be returned by `findNextAppropriateProvider`

I guess that would solve the issue, wouldn't it? Could you give it a try?

**Comment by secaw** [comment]
I also got a similar ANR, when zooming far out on a tablet with several providers active. Note that when zooming far out on a 10'' tablet with your fingers without releasing (3-4 zoom levels), a few hundred tiles may be loaded at once, so the `tileFileSystemMaxQueueSize` is easily being reached. Tile loading on some providers may take some time, especially on one like mapsforge. When this is the case due to the logic of `removeEldestEntry` -> `getCallback().mapTileRequestFailed`, currently the provider chain of `MapTileProviderArray` is kept being worked through, thus a lot of work pending!
Personally I don't see any reason, why the next provider is being tried for a tile that should be skipped for now due to exceeded MaxQueueSize. In my case, besides the extreme CPU load and ANR this actually resulted in a strange patchwork of tiles from different sources.

Related: #496 

I have this solution, which for me works really well and I think, it might be a general fix:

1 - In `IMapTileProviderCallback` I have a new entry underneeth `mapTileRequestFailed`:
```
/**
* The map tile request has failed - exceeds MaxQueueSize.
*
* @param aState
*            a state object
*/
void mapTileRequestFailedExceedsMaxQueueSize(MapTileRequestState aState);
```
2 - In `MapTileModuleProviderBase` -> `removeEldestEntry` at the bottom, I changed
```
state.getCallback().mapTileRequestFailed(state);
```
to
```
state.getCallback().mapTileRequestFailedExceedsMaxQueueSize(state);
```
3 - In `MapTileProviderArray` underneeth `mapTileRequestFailed` I added
```	
@Override
public void mapTileRequestFailedExceedsMaxQueueSize(final MapTileRequestState aState) {
	mapTileRequestFailed(aState);
}
```
4 - Finally in my custom `MapTileProviderArray` I added
```
@Override
public void mapTileRequestFailedExceedsMaxQueueSize(final MapTileRequestState aState) {
    synchronized (mWorking) {
        mWorking.remove(aState.getMapTile());
    }
    super.mapTileRequestFailed(aState);
}
```
What do you think?

Besides that @monsieurtanuki , would it be at all possible to change the zooming out tile loading that way that it dynamically switches zoom levels, while zooming out, so that the loading of a few hundred tiles at once as described above doesn't happen anymore?

**Comment by monsieurtanuki** [comment]
@secaw If we remove the scale from MapView as expressed in #790, we'll try to get the tiles of the most relevant zoom at any moment.
Today, when you zoom out 4 levels with pinch gesture and hold, you get a tiny (divided by 4) version of the tiles around the center, that is, not only you try to download tiles that are geographically too remote, but also you display them with a too small scale. #790 will download only what is relevant for the current zoom (that will include the scale).
We should also be able to remove download queue items when they don't belong to the current zoom. But that optimization is beyond #790.

**Comment by monsieurtanuki** [comment]
Little precision/correction: for a 4x zoom out, I think we currently try to get 2^4 more tiles than needed (and display them 2^4 smaller). Than would explain the hundreds of downloaded tiles.

**Comment by secaw** [comment]
@monsieurtanuki Yes exactly. I actually thought, that #790 might solve some of these problems when I read it, but have no idea of the scaling mechanisms myself. Suggested improvement in #790 sounds fantastic overall.
I still think, my above described fix is very helpful in giving us control over what happens to tiles exceeding `MaxQueueSize`.

**Comment by monsieurtanuki** [comment]
> I still think, my above described fix is very helpful in giving us control over what happens to tiles exceeding MaxQueueSize.

@secaw My recent comments were mostly focused on a better control over the requested zoom levels. Obviously something has also to be fixed when the max queue size is exceeded, and your solution looks good (disclaimer: I didn't delve into the code and therefore didn't check the potential side effects of removing the tile from the queue before calling `mapTileRequestFailed` on the very same tile).

**Comment by monsieurtanuki** [comment]
@Vandalko In addition to what @secaw developed and merged regarding the tile queue side effects, I just merged in #800 a code that does not try to download dozens of tiles during the zoom. Could you give it a try?

**Comment by monsieurtanuki** [comment]
Fixed.
@Vandalko Feel free to reopen if needed.
